pkg/operator/operator.go

```diff
@@ -241,76 +273,237 @@ func (k *Kontroller) withLeaderElection() error {
+	}
+
+	preRebootNodes := k8sutil.FilterNodesByRequirement(nodelist.Items, beforeRebootReq)
+
+	for _, n := range preRebootNodes {
+		if hasAllAnnotations(n, k.beforeRebootAnnotations) {
+			glog.V(4).Infof("Deleting label %q for %q", constants.LabelBeforeReboot, n.Name)
+			glog.V(4).Infof("Setting annotation %q to true for %q", constants.AnnotationOkToReboot, n.Name)
+			err = k8sutil.UpdateNodeRetry(k.nc, n.Name, func(node *v1api.Node) {
+				delete(node.Labels, constants.LabelBeforeReboot)
+
+				// make sure that the node still wants to reboot
+				if !wantsRebootSelector.Matches(fields.Set(node.Annotations)) {
+					glog.Warningf("Node %v no longer wanted to reboot while we were trying to mark it so: %v", node.Name, node.Annotations)
+					return
+				}
+
+				node.Annotations[constants.AnnotationOkToReboot] = constants.True
+			})
+			if err != nil {
+				return fmt.Errorf("Failed to update node %q: %v", n.Name, err)
+			}
 		}
 	}
 
-	nodelist, err = k.nc.List(v1meta.ListOptions{})
+	return nil
+}
+
+// checkAfterReboot gets all nodes with the after-reboot=true label and checks
+// if  all of the configured after-reboot annotations are set to true. If they
+// are, it deletes the after-reboot=true label and sets reboot-ok=false to tell
+// the agent that it has completed it's reboot successfully.
+// If there is an error getting the list of nodes or updating any of them, an
+// error is immediately returned.
+func (k *Kontroller) checkAfterReboot() error {
+	nodelist, err := k.nc.List(v1meta.ListOptions{})
 	if err != nil {
-		glog.Infof("Failed listing nodes: %v", err)
-		return
+		return fmt.Errorf("Failed listing nodes: %v", err)
+	}
+
+	postRebootNodes := k8sutil.FilterNodesByRequirement(nodelist.Items, afterRebootReq)
+
+	for _, n := range postRebootNodes {
+		if hasAllAnnotations(n, k.afterRebootAnnotations) {
+			glog.V(4).Infof("Deleting label %q for %q", constants.LabelAfterReboot, n.Name)
+			glog.V(4).Infof("Setting annotation %q to false for %q", constants.AnnotationOkToReboot, n.Name)
+			err = k8sutil.UpdateNodeRetry(k.nc, n.Name, func(node *v1api.Node) {
+				delete(node.Labels, constants.LabelAfterReboot)
+				node.Annotations[constants.AnnotationOkToReboot] = constants.False
+			})
+			if err != nil {
+				return fmt.Errorf("Failed to update node %q: %v", n.Name, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// markBeforeReboot gets nodes which want to reboot and marks them with the
+// before-reboot=true label. This is considered the beginning of the reboot
+// process from the perspective of the update-operator. It will only mark
+// nodes with this label up to the maximum number of concurrently rebootable
+// nodes as configured with the maxRebootingNodes constant.
+// If there is an error getting the list of nodes or updating any of them, an
+// error is immediately returned.
+func (k *Kontroller) markBeforeReboot() error {
+	nodelist, err := k.nc.List(v1meta.ListOptions{})
```

Comment:
I have some ideas around that, but I think that for now we should just keep the list call in this function and maybe change it later.